Render interactive form (AcroForm) text widget annotations

timvandermeij opened the pull request:
This patch is the first step towards implementing support for interactive forms (AcroForms). It makes it possible to render text widget annotations exactly like Adobe Reader/Acrobat.

Everything we implement for AcroForms is disabled by default using a preference, mainly because it is not ready to use yet, but has to implemented in many steps to avoid complexity. The preference allows us to work with the code while not exposing the behavior by default. Mainly storing entered values and printing them is still absent, which would be minimal requirements for enabling this by default. The default behavior is therefore unchanged.

You can test the functionality using http://web.archive.org/web/20110918100215/http://www.irs.gov/pub/irs-pdf/f1040.pdf (from https://github.com/mozilla/pdf.js/blob/master/test/pdfs/f1040.pdf.link) and setting the preference `renderInteractiveForms` to `true`.

@Snuffleupagus Could you review this PR? I talked to Yury on IRC and he agreed that the preference approach is most useful here so we can iterate on this.

_Easier reviewing with https://github.com/mozilla/pdf.js/pull/7602/files?w=1._

timvandermeij committed:
Render interactive form (AcroForm) text widget annotations

This patch is the first step towards implementing support for
interactive forms (AcroForms). It makes it possible to render text
widget annotations exactly like Adobe Reader/Acrobat.

Everything we implement for AcroForms is disabled by default using a
preference, mainly because it is not ready to use yet, but has to
implemented in many steps to avoid complexity. The preference allows us
to work with the code while not exposing the behavior by default. Mainly
storing entered values and printing them is still absent, which would be
minimal requirements for enabling this by default.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented:
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR36kC_Mrze9CACrQRI:-KR36kCal4qQEiBbSFt4:bxqe87y) (no related file):*
First, let me just say that it's *really* nice that you've started working on support for forms, and even nicer of you to do it in such manageable chunks :-)

Besides the inline comments, pretty *please* add reference (and if applicable perhaps also unit) tests!
Since you're adding forms support in this incremental manner, it should hopefully not be too difficult to add tests along the way to ensure that we get at least decent test-coverage from the start.

My suggestion is to add a `forms` flag to the `eq` type reference tests, that takes the same code-path in `test/driver` as the existing `annotations` flag. The file `f1040.pdf` is already present in the test-suite as `load` test, so you should be able to use that one.

---

*[src/display/annotation_layer.js, line 442 \[r1\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR35BDhoUJ6nQpynRlL:-KR35BDhoUJ6nQpynRlM:bz8hn8y) ([raw file](https://github.com/mozilla/pdf.js/blob/3d8e07151dd242f0acdb893d6311d8b9f77b8e31/src/display/annotation_layer.js#L442)):*
> ```JavaScript
>       this.container.className = 'textWidgetAnnotation';
> 
>       var renderInteractiveForms = getDefaultSetting('renderInteractiveForms');
> ```

Let's pass in a `renderInteractiveForms` parameter when creating the annotationLayer in `web/annotation_layer_builder.js` instead. I.e. for now I think it'd be sufficient to just add a
`renderInteractiveForms: PDFJS.renderInteractiveForms,` entry to https://github.com/mozilla/pdf.js/blob/master/web/annotation_layer_builder.js#L75 and then access that here instead.

(We can always improve on these things later on, e.g. how the preference is passed along to the annotationLayer, when we get closer to enabling annotations by default.)

---

*[src/display/dom_utils.js, line 213 \[r1\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR348kEQjG2FT56giGj:-KR348kFeExAWkK6ByXa:bxc5wlk) ([raw file](https://github.com/mozilla/pdf.js/blob/3d8e07151dd242f0acdb893d6311d8b9f77b8e31/src/display/dom_utils.js#L213)):*
> ```JavaScript
>       return !!(globalSettings && globalSettings.enableStats);
>     case 'renderInteractiveForms':
>       return globalSettings ? globalSettings.renderInteractiveForms : false;
> ```

Please note the comment at the [top of this list](https://github.com/mozilla/pdf.js/blob/master/src/display/dom_utils.js#L157-L158):
```javascript
// The list of the settings and their default is maintained for backward
// compatibility and shall not be extended or modified. See also global.js.
```
Since this is a *new* setting, it shouldn't be added here. Please see the other comments for another approach.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7602)*
<!-- Sent from Reviewable.io -->

timvandermeij commented:
Thank you for the review! I had planned to implement unit/reference testing in the next PR, but I can see how it will already be useful here and hopefully it should not take too much code to implement, so I'll instead do that for this PR as you suggested.

Snuffleupagus commented:
Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions.

---

*[web/annotation_layer_builder.js, line 83 \[r2\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR3p2KdXIvQDP-FOQK5:-KR3p2KdXIvQDP-FOQK6:brn8f8i) ([raw file](https://github.com/mozilla/pdf.js/blob/be6838d4c9327a658dd0d553281594b85ece3457/web/annotation_layer_builder.js#L83)):*
> ```JavaScript
>           linkService: self.linkService,
>           downloadManager: self.downloadManager,
>           renderInteractiveForms: PDFJS.renderInteractiveForms,
> ```

Please use `pdfjsLib` instead of directly accessing `PDFJS` here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7602)*
<!-- Sent from Reviewable.io -->

timvandermeij committed:
Enable regression testing for interactive forms

timvandermeij commented:
I have addressed the comments. Reference testing has been enabled in a new commit. Locally this can be tested by removing all entries from `test/test_manifest.json` except for the newly added `f1040-eq` test and running `gulp botmakeref`. Be sure to back-up your current reference images just in case. After this, inspect the reference images to see the blue input fields.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

Snuffleupagus commented:
Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[test/test_manifest.json, line 735 \[r4\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR46Yv3N8HxVRtYQDud:-KR46Yv3N8HxVRtYQDue:b-pa6azn) ([raw file](https://github.com/mozilla/pdf.js/blob/18cf425f0c41b7070a04b5d8ea4db01366ecc407/test/test_manifest.json#L735)):*
> ```JSON
>        "type": "load"
>     },
>     {  "id": "f1040-eq",
> ```

Let's make this `f1040-forms` instead.

---

*[test/test_manifest.json, line 741 \[r4\]](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR46e1NinTT1XZrAfBd:-KR46e1NinTT1XZrAfBe:b-tfr209) ([raw file](https://github.com/mozilla/pdf.js/blob/18cf425f0c41b7070a04b5d8ea4db01366ecc407/test/test_manifest.json#L741)):*
> ```JSON
>        "rounds": 1,
>        "type": "eq",
>        "annotations": true,
> ```

Would it be possible to implement the `forms` option in such a way that it's not necessary to also specify `annotations`, e..g with `if (task.annotations || task.forms) {` in https://github.com/mozilla/pdf.js/blob/8dbb5a7c4a63782e24382f9773386d9facbfc557/test/driver.js#L486?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7602)*
<!-- Sent from Reviewable.io -->

timvandermeij commented:
The review comments have been addressed, so let's re-run the tests to be sure that the changes are fine.

/botio test

Snuffleupagus commented:
Reviewed 5 of 7 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 2 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/mozilla/pdf.js/7602#-KR7qbDQmAbUJVv_oraa:-KR7qbDR6gf-wLY9si_5:bm6wfmv) (no related file):*
A nice start on forms support, thanks for doing this and especially for adding test!

/botio makeref

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/pdf.js/7602)*
<!-- Sent from Reviewable.io -->
